README.md
# NAME

Mojolicious::Plugin::WebPush - plugin to aid real-time web push

# SYNOPSIS

    # Mojolicious::Lite
    my $sw = plugin 'ServiceWorker' => { debug => 1 };
    my $webpush = plugin 'WebPush' => {
      save_endpoint => '/api/savesubs',
      subs_session2user_p => \&subs_session2user_p,
      subs_create_p => \&subs_create_p,
      subs_read_p => \&subs_read_p,
      subs_delete_p => \&subs_delete_p,
      ecc_private_key => 'vapid_private_key.pem',
      claim_sub => "mailto:admin@example.com",
    };

    sub subs_session2user_p {
      my ($session) = @_;
      return Mojo::Promise->reject("Session not logged in") if !$session->{user_id};
      Mojo::Promise->resolve($session->{user_id});
    }

    sub subs_create_p {
      my ($session, $subs_info) = @_;
      app->db->save_subs_p($session->{user_id}, $subs_info);
    }

    sub subs_read_p {
      my ($user_id) = @_;
      app->db->lookup_subs_p($user_id);
    }

    sub subs_delete_p {
      my ($user_id) = @_;
      app->db->delete_subs_p($user_id);
    }

# DESCRIPTION

[Mojolicious::Plugin::WebPush](https://metacpan.org/pod/Mojolicious::Plugin::WebPush) is a [Mojolicious](https://metacpan.org/pod/Mojolicious) plugin. In
order to function, your app needs to have first installed
[Mojolicious::Plugin::ServiceWorker](https://metacpan.org/pod/Mojolicious::Plugin::ServiceWorker) as shown in the synopsis above.

# METHODS

[Mojolicious::Plugin::WebPush](https://metacpan.org/pod/Mojolicious::Plugin::WebPush) inherits all methods from
[Mojolicious::Plugin](https://metacpan.org/pod/Mojolicious::Plugin) and implements the following new ones.

## register

    my $p = $plugin->register(Mojolicious->new, \%conf);

Register plugin in [Mojolicious](https://metacpan.org/pod/Mojolicious) application, returning the plugin
object. Takes a hash-ref as configuration, see ["OPTIONS"](#options) for keys.

# OPTIONS

## save\_endpoint

Required. The route to be added to the app for the service worker to
register users for push notification. The handler for that will call
the ["subs\_create\_p"](#subs_create_p). If success is indicated, it will return JSON:

    { "data": { "success": true } }

If failure:

    { "errors": [ { "message": "The exception reason" } ] }

This will be handled by the provided service worker. In case it is
required by the app itself, the added route is named `webpush.save`.

## subs\_session2user\_p

Required. The code to be called to look up the user currently identified
by this session, which returns a promise of the user ID. Must reject
if no user logged in and that matters. It will be passed parameters:

- The ["session" in Mojolicious::Controller](https://metacpan.org/pod/Mojolicious::Controller#session) object, to correctly identify
the user.

## subs\_create\_p

Required. The code to be called to store users registered for push
notifications, which must return a promise of a true value if the
operation succeeds, or reject with a reason. It will be passed parameters:

- The ID to correctly identify the user. Please note that you ought to
allow one person to have several devices with web-push enabled, and to
design accordingly.
- The `subscription_info` hash-ref, needed to push actual messages.

## subs\_read\_p

Required. The code to be called to look up a user registered for push
notifications. It will be passed parameters:

- The opaque information your app uses to identify the user.

Returns a promise of the `subscription_info` hash-ref. Must reject if
not found.

## subs\_delete\_p

Required. The code to be called to delete up a user registered for push
notifications. It will be passed parameters:

- The opaque information your app uses to identify the user.

Returns a promise of the deletion result. Must reject if not found.

## ecc\_private\_key

A value to be passed to ["new" in Crypt::PK::ECC](https://metacpan.org/pod/Crypt::PK::ECC#new): a simple scalar is a
filename, a scalar-ref is the actual key. If not provided,
["webpush.authorization"](#webpush-authorization) will (obviously) not be able to function.

## claim\_sub

A value to be used as the `sub` claim by the ["webpush.authorization"](#webpush-authorization),
which needs it. Must be either an HTTPS or `mailto:` URL.

## claim\_exp\_offset

A value to be added to current time, in seconds, in the `exp` claim
for ["webpush.authorization"](#webpush-authorization). Defaults to 86400 (24 hours). The maximum
valid value in RFC 8292 is 86400.

## push\_handler

Override the default push-event handler supplied to
["add\_event\_listener" in Mojolicious::Plugin::ServiceWorker](https://metacpan.org/pod/Mojolicious::Plugin::ServiceWorker#add_event_listener). The default
will interpret the message as a JSON object. The key `title` will be
the notification title, deleted from that object, then the object will be
the options passed to `<ServiceWorkerRegistration>.showNotification`.

See
[https://developers.google.com/web/fundamentals/push-notifications/handling-messages](https://developers.google.com/web/fundamentals/push-notifications/handling-messages)
for possibilities.

# HELPERS

## webpush.create\_p

    $c->webpush->create_p($user_id, $subs_info)->then(sub {
      $c->render(json => { data => { success => \1 } });
    });

## webpush.read\_p

    $c->webpush->read_p($user_id)->then(sub {
      $c->render(text => 'Info: ' . to_json(shift));
    });

## webpush.delete\_p

    $c->webpush->delete_p($user_id)->then(sub {
      $c->render(json => { data => { success => \1 } });
    });

## webpush.authorization

    my $header_value = $c->webpush->authorization;

Won't function without ["claim\_sub"](#claim_sub) and ["ecc\_private\_key"](#ecc_private_key). Returns
a suitable `Authorization` header value to send to a push service.
Valid for a period defined by ["claim\_exp\_offset"](#claim_exp_offset). Not currently cached,
but could become so to avoid unnecessary computation.

## webpush.aud

    my $aud = $c->webpush->aud;

Gives the app's value it will use for the `aud` JWT claim, useful mostly
for testing.

## webpush.public\_key

    my $pkey = $c->webpush->public_key;

Gives the app's public VAPID key, calculated from the private key.

## webpush.verify\_token

    my $bool = $c->webpush->verify_token($authorization_header_value);

Cryptographically verifies a JSON Web Token (JWT), such as generated
by ["webpush.authorization"](#webpush-authorization).

## webpush.encrypt

    use MIME::Base64 qw(decode_base64url);
    my $ciphertext = $c->webpush->encrypt($data_bytes,
      map decode_base64url($_), @{$subscription_info->{keys}}{qw(p256dh auth)}
    );

Returns the data encrypted according to RFC 8188, for the relevant
subscriber.

# TEMPLATES

Various templates are available for including in the app's templates:

## webpush-askPermission.html.ep

JavaScript functions, also for putting inside a `script` element:

- askPermission
- subscribeUserToPush
- sendSubscriptionToBackEnd

These each return a promise, and should be chained together:

    <button onclick="
      askPermission().then(subscribeUserToPush).then(sendSubscriptionToBackEnd)
    ">
      Ask permission
    </button>
    <script>
    %= include 'serviceworker-install'
    %= include 'webpush-askPermission'
    </script>

Each application must decide when to ask such permission, bearing in
mind that once permission is refused, it is very difficult for the user
to change such a refusal.

When it is granted, the JavaScript code will communicate with the
application, registering the needed information needed to web-push.

# SEE ALSO

[Mojolicious](https://metacpan.org/pod/Mojolicious), [Mojolicious::Guides](https://metacpan.org/pod/Mojolicious::Guides), [https://mojolicious.org](https://mojolicious.org).

[Mojolicious::Command::webpush](https://metacpan.org/pod/Mojolicious::Command::webpush) - command-line control of web-push.

RFC 8292 - Voluntary Application Server Identification (for web push).

[Crypt::RFC8188](https://metacpan.org/pod/Crypt::RFC8188) - Encrypted Content-Encoding for HTTP (using `aes128gcm`).

[https://developers.google.com/web/fundamentals/push-notifications](https://developers.google.com/web/fundamentals/push-notifications)

# ACKNOWLEDGEMENTS

Part of this code is ported from
[https://github.com/web-push-libs/pywebpush](https://github.com/web-push-libs/pywebpush).


Changes
- update example to work right with $c in signature
- zap webpush.aud as is supposed to be base URL of push service, not of app

0.04 2021-02-10
- add $c to all callbacks - thanks @uniejo

0.03 2020-09-05
- (hopefully) fix JSON error in tests

0.02 2020-07-20
- make example app handle empty file on keygen

0.01 2020-07-20
- first version


t/lib/TestUtils.pm
package TestUtils;

use Mojo::Promise;
use Exporter 'import';

our @EXPORT_OK = qw(webpush_config $ENDPOINT %userdb);

our $ENDPOINT = '/api/savesubs';
our %userdb;

sub webpush_config {
  +{
    save_endpoint => $ENDPOINT,
    subs_session2user_p => \&subs_session2user_p,
    subs_create_p => \&subs_create_p,
    subs_read_p => \&subs_read_p,
    subs_delete_p => \&subs_delete_p,
  };
}

sub subs_session2user_p {
  return Mojo::Promise->reject("Session not logged in") if !$_[0]{user_id};
  Mojo::Promise->resolve($_[0]{user_id});
}

sub subs_create_p {
  my ($user_id, $subs_info) = @_;
  $userdb{$user_id} = $subs_info;
  Mojo::Promise->resolve(1);
}

sub subs_read_p {
  my ($user_id) = @_;
  return Mojo::Promise->reject("Not found: '$user_id'") if !$userdb{$user_id};
  Mojo::Promise->resolve($userdb{$user_id});
}

sub subs_delete_p {
  my ($user_id) = @_;
  return Mojo::Promise->reject("Not found: '$user_id'") if !$userdb{$user_id};
  Mojo::Promise->resolve(delete $userdb{$user_id});
}

1;


Makefile.PL
use 5.008001;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
  NAME         => 'Mojolicious::Plugin::WebPush',
  VERSION_FROM => 'lib/Mojolicious/Plugin/WebPush.pm',
  ABSTRACT_FROM => 'lib/Mojolicious/Plugin/WebPush.pm',
  AUTHOR           => q{Ed J <etj@cpan.org>},
  LICENSE          => 'artistic_2',
  MIN_PERL_VERSION => '5.014',
  CONFIGURE_REQUIRES => {
    'ExtUtils::MakeMaker' => '7.10',
  },
  TEST_REQUIRES => {
    'Test::More' => '0.98',
  },
  PREREQ_PM => {
    'Mojolicious' => '8.42',
    'Crypt::JWT' => '0.026',
    'CryptX' => '0.068',
    'Crypt::RFC8188' => '0.01',
  },
  clean => { FILES => 'Mojolicious-Plugin-WebPush-*' },
  META_MERGE => {
    "meta-spec" => { version => 2 },
    dynamic_config => 0,
    resources => {
      x_IRC => 'irc://irc.perl.org/#graphql-perl',
      repository => {
        type => 'git',
        url => 'git@github.com:mohawk2/Mojolicious-Plugin-WebPush.git',
        web => 'https://github.com/mohawk2/Mojolicious-Plugin-WebPush',
      },
      bugtracker  => {
        web => 'https://github.com/mohawk2/Mojolicious-Plugin-WebPush/issues',
      },
      license => [ 'http://dev.perl.org/licenses/' ],
    },
    prereqs => {
      develop => {
        requires => {
          'Test::Pod' => '1.22',
          'Pod::Markdown' => 0,
        },
      },
    },
  },
);

sub MY::postamble {
  <<EOF;
pure_all :: README.md

README.md : \$(VERSION_FROM)
\tpod2markdown \$< >\$\@
EOF
}


examples/app.pl
use Mojolicious::Lite;
use Mojo::File qw(curfile path);
use Mojo::Promise;
use Mojo::SQLite;
use Mojo::JSON qw(decode_json encode_json);

my $pushkeyfile = curfile->sibling('webpush_private_key.pem')
  ->to_rel(path)->to_string;
if (! -f $pushkeyfile) {
  die <<EOF;
No webpush private key file found in '$pushkeyfile'. Make one:
  $0 webpush keygen >$pushkeyfile
EOF
}

helper sqlite => sub {
  state $path = app->home->child( 'data.db' );
  state $sqlite = Mojo::SQLite->new( 'sqlite:' . $path );
  return $sqlite;
};
app->sqlite->auto_migrate(1)->migrations->from_data;

my $sw = plugin 'ServiceWorker' => {
  debug => 1,
  network_first => [
    "whoami",
  ],
  network_only => [
    "push", "api/savesubs",
  ],
};
my $webpush = plugin 'WebPush' => webpush_config();

get '/' => 'index';

post '/login' => sub {
  my $user_id = $_[0]->req->json('/user_id');
  $_[0]->session(user_id => $user_id);
  $_[0]->render(json => { data => { user_id => $user_id } });
}, 'login';

post '/push' => sub {
  my ($c) = @_;
  my $user_id = $c->session->{user_id};
  my $message = $c->req->json;
  $c->webpush->send_p($message, $user_id, 30, 'normal')->then(sub {
    $c->render(json => $_[0]);
  });
}, 'push';

get '/whoami' => sub {
  $_[0]->render(json => { user_id => $_[0]->session('user_id') });
}, 'whoami';

sub webpush_config {
  +{
    save_endpoint => '/api/savesubs',
    subs_session2user_p => \&subs_session2user_p,
    subs_create_p => \&subs_create_p,
    subs_read_p => \&subs_read_p,
    subs_delete_p => \&subs_delete_p,
    ecc_private_key => $pushkeyfile,
    claim_sub => 'mailto:admin@example.com',
  };
}

sub subs_session2user_p {
  return Mojo::Promise->reject("Session not logged in") if !$_[0]{user_id};
  Mojo::Promise->resolve($_[0]{user_id});
}

sub subs_create_p {
  my ($user_id, $subs_info) = @_;
  app->sqlite->db->insert(
    'users',
    { username => $user_id, subs_info => encode_json($subs_info) },
  );
  Mojo::Promise->resolve(1);
}

sub subs_read_p {
  my ($user_id) = @_;
  my @results = app->sqlite->db->select(
    'users',
    [qw(subs_info)],
    { username => $user_id },
  )->arrays->each;
  return Mojo::Promise->reject("Not found: '$user_id'") if !@results;
  Mojo::Promise->resolve(decode_json $results[0][0]);
}

sub subs_delete_p {
  my ($user_id) = @_;
  return app->webpush->read_p($user_id)->then(sub {
    app->sqlite->db->delete('users', { username => $user_id });
    $_[0];
  });
}

app->start;

__DATA__

@@ index.html.ep
<div>You are logged in as: <span id="log"><i>not logged in</i></span></div>
<form onsubmit="login(this.children[1].value); return false">
<input type="submit" value="Login as"><input value="bob">
</form>
<p/>
<button onclick="askPermission().then(subscribeUserToPush).then(sendSubscriptionToBackEnd)">Ask permission</button>
<p/>
<form onsubmit="push(this.children[1].value); return false">
<input type="submit" value="Push message"><input value="Test push message from sample"><span id="push_status"><i>no push yet</i></span>
</form>
<script>
%= include 'serviceworker-install'
%= include 'webpush-askPermission'
var login_disp = document.getElementById('log');
function set_logged_in(u) { login_disp.innerHTML = u }
var push_status = document.getElementById('push_status');
function set_push_status(m) { push_status.innerHTML = m }
const whoami = () => fetch('whoami', { credentials: 'include' }).then(
  r => r.json()).then(j => j.user_id);
const showwhoiam = () => whoami().then(set_logged_in);
showwhoiam();
function json_fetch(url, data) {
  return fetch(url, {
    method: 'POST',
    credentials: 'include',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  }).then(
    r => r.json()
  ).then(
    j => j.errors ? Promise.reject(j.errors[0].message) : j.data
  );
}
function login(newname) {
  return json_fetch('login', { user_id: newname }).then(
    showwhoiam,
    e => set_logged_in('<i>' + e + '</i>'),
  );
}
function push(message) {
  return json_fetch('push', { title: message, body: message }).then(
    () => set_push_status('Success'),
    e => set_push_status('<i>Push failed: ' + e + '</i>'),
  );
}
</script>

@@ migrations
-- 1 up
CREATE TABLE users (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    username VARCHAR(100) UNIQUE NOT NULL,
    subs_info VARCHAR(255) NOT NULL
);
-- 1 down
DROP TABLE users;


lib/Mojolicious/Command/webpush.pm
package Mojolicious::Command::webpush;
use Mojo::Base 'Mojolicious::Command';
use Mojo::JSON qw(encode_json decode_json);
use Crypt::PK::ECC;

my %COMMAND2JSON = (
  create => [ 1 ],
);
my %COMMAND2CB = (
  keygen => \&_keygen,
);

has description => q{Manage your app's web-push};
has usage       => sub { shift->extract_usage };

sub _keygen {
  print Crypt::PK::ECC->new->generate_key('prime256v1')
    ->export_key_pem('private');
}

sub _promisify {
  my ($app, $cmd) = @_;
  sub {
    my (undef, @args) = @_;
    my @res = eval { $app->$cmd(@args) };
    $@ ? Mojo::Promise->reject($@) : Mojo::Promise->resolve(@res);
  };
}

sub run {
  my ($self, $cmd, @args) = @_;
  return $COMMAND2CB{$cmd}->($self, @args) if $COMMAND2CB{$cmd};
  $args[$_] = decode_json($args[$_]) for @{ $COMMAND2JSON{$cmd} || [] };
  $cmd .= "_p";
  $self->app->webpush->$cmd(@args)->then(
    sub { print encode_json(@_), "\n" },
    sub { print STDERR @_, "\n" },
  )->wait;
}

1;

=encoding utf8

=head1 NAME

Mojolicious::Command::webpush - Manage your app's web-push

=head1 SYNOPSIS

  Usage: APPLICATION webpush COMMAND [OPTIONS]

    ./myapp.pl webpush create <USERID> <JSON>
    ./myapp.pl webpush read <USERID>
    ./myapp.pl webpush delete <USERID>
    ./myapp.pl webpush keygen > webpush_private_key.pem

  Options:
    -h, --help          Show this summary of available options
        --home <path>   Path to home directory of your application, defaults to
                        the value of MOJO_HOME or auto-detection
    -m, --mode <name>   Operating mode for your application, defaults to the
                        value of MOJO_MODE/PLACK_ENV or "development"

=head1 DESCRIPTION

L<Mojolicious::Command::webpush> manages your application's web-push
information. It gives a command-line interface to the relevant helpers
in L<Mojolicious::Plugin::WebPush/HELPERS>.

The C<keygen> command prints a PEM-encoded L<Crypt::PK::ECC/generate_key>
C<prime256v1> result.

=head1 ATTRIBUTES

L<Mojolicious::Command::webpush> inherits all attributes from
L<Mojolicious::Command> and implements the following new ones.

=head1 METHODS

L<Mojolicious::Command::webpush> inherits all methods from
L<Mojolicious::Command> and implements the following new ones.

=head1 SEE ALSO

L<Mojolicious::Plugin::WebPush>

=cut


lib/Mojolicious/Plugin/WebPush.pm
package Mojolicious::Plugin::WebPush;
use Mojo::Base 'Mojolicious::Plugin';
use Mojo::JSON qw(decode_json encode_json);
use Mojo::URL;
use Crypt::PK::ECC;
use MIME::Base64 qw(encode_base64url decode_base64url);
use Crypt::JWT qw(encode_jwt decode_jwt);
use Crypt::RFC8188 qw(ece_encrypt_aes128gcm);

our $VERSION = '0.04';

my @MANDATORY_CONF = qw(
  subs_session2user_p
  save_endpoint
  subs_create_p
  subs_read_p
  subs_delete_p
);
my @AUTH_CONF = qw(claim_sub ecc_private_key);
my $DEFAULT_PUSH_HANDLER = <<'EOF';
event => {
  var msg = event.data.json();
  var title = msg.title;
  delete msg.title;
  event.waitUntil(self.registration.showNotification(title, msg));
}
EOF

sub _decode {
  my ($bytes) = @_;
  my $body = eval { decode_json($bytes) };
  # conceal error info like versions from attackers
  return (0, "Malformed request") if $@;
  (1, $body);
}

sub _error {
  my ($c, $error) = @_;
  $c->render(status => 500, json => { errors => [ { message => $error } ] });
}

sub _make_route_handler {
  my ($subs_session2user_p, $subs_create_p) = @_;
  sub {
    my ($c) = @_;
    my ($decode_ok, $body) = _decode($c->req->body);
    return _error($c, $body) if !$decode_ok;
    eval { validate_subs_info($body) };
    return _error($c, $@) if $@;
    return $subs_session2user_p->($c, $c->session)->then(
      sub { $subs_create_p->($c, $_[0], $body) },
    )->then(
      sub { $c->render(json => { data => { success => \1 } }) },
      sub { _error($c, @_) },
    );
  };
}

sub _make_auth_helper {
  my ($app, $conf) = @_;
  my $exp_offset = $conf->{claim_exp_offset} || 86400;
  my $key = Crypt::PK::ECC->new($conf->{ecc_private_key});
  my $claims_start = { sub => $conf->{claim_sub} };
  my $pkey = encode_base64url $key->export_key_raw('public');
  $app->helper('webpush.public_key' => sub { $pkey });
  sub {
    my ($c, $subs_info) = @_;
    my $aud = Mojo::URL->new($subs_info->{endpoint})->path(Mojo::Path->new->trailing_slash(0)).'';
    my $claims = { aud => $aud, exp => time + $exp_offset, %$claims_start };
    my $token = encode_jwt key => $key, alg => 'ES256', payload => $claims;
    "vapid t=$token,k=$pkey";
  };
}

sub _verify_helper {
  my ($app, $auth_header_value) = @_;
  (my $schema, $auth_header_value) = split ' ', $auth_header_value;
  return if $schema ne 'vapid';
  my %k2v = map split('=', $_), split ',', $auth_header_value;
  eval {
    my $key = Crypt::PK::ECC->new;
    $key->import_key_raw(decode_base64url($k2v{k}), 'P-256');
    decode_jwt token => $k2v{t}, key => $key, alg => 'ES256', verify_exp => 0;
  };
}

sub _encrypt_helper {
  my ($c, $plaintext, $receiver_key, $auth_key) = @_;
  die "Invalid p256dh key specified\n"
    if length($receiver_key) != 65 or $receiver_key !~ /^\x04/;
  my $onetime_key = Crypt::PK::ECC->new->generate_key('prime256v1');
  ece_encrypt_aes128gcm(
    $plaintext, (undef) x 2, $onetime_key, $receiver_key, $auth_key,
  );
}

sub _send_helper {
  my ($c, $message, $user_id, $ttl, $urgency) = @_;
  $ttl ||= 30;
  $urgency ||= 'normal';
  $c->webpush->read_p($user_id)->then(sub {
    my ($subs_info) = @_;
    my $body = $c->webpush->encrypt(
      encode_json($message),
      map decode_base64url($_), @{$subs_info->{keys}}{qw(p256dh auth)}
    );
    my $headers = {
      Authorization => $c->webpush->authorization($subs_info),
      'Content-Length' => length($body),
      'Content-Encoding' => 'aes128gcm',
      TTL => $ttl,
      Urgency => $urgency,
    };
    $c->app->ua->post_p($subs_info->{endpoint}, $headers, $body);
  })->then(sub {
    my ($tx) = @_;
    return $c->webpush->delete_p($user_id)->then(sub {
      { data => { success => \1 } }
    }) if $tx->res->code == 404 or $tx->res->code == 410;
    return { errors => [ { message => $tx->res->body } ] }
      if $tx->res->code > 399;
    { data => { success => \1 } };
  }, sub {
    { errors => [ { message => $_[0] } ] }
  });
}

sub register {
  my ($self, $app, $conf) = @_;
  my @config_errors = grep !exists $conf->{$_}, @MANDATORY_CONF;
  die "Missing config keys @config_errors\n" if @config_errors;
  $app->helper('webpush.create_p' => sub {
    eval { validate_subs_info($_[2]) };
    return Mojo::Promise->reject($@) if $@;
    goto &{ $conf->{subs_create_p} };
  });
  $app->helper('webpush.read_p' => $conf->{subs_read_p});
  $app->helper('webpush.delete_p' => $conf->{subs_delete_p});
  $app->helper('webpush.authorization' => (grep !$conf->{$_}, @AUTH_CONF)
    ? sub { die "Must provide @AUTH_CONF\n" }
    : _make_auth_helper($app, $conf)
  );
  $app->helper('webpush.verify_token' => \&_verify_helper);
  $app->helper('webpush.encrypt' => \&_encrypt_helper);
  $app->helper('webpush.send_p' => \&_send_helper);
  my $r = $app->routes;
  $r->post($conf->{save_endpoint} => _make_route_handler(
    @$conf{qw(subs_session2user_p subs_create_p)},
  ), 'webpush.save');
  push @{ $app->renderer->classes }, __PACKAGE__;
  $app->serviceworker->add_event_listener(
    push => $conf->{push_handler} || $DEFAULT_PUSH_HANDLER
  );
  $self;
}

sub validate_subs_info {
  my ($info) = @_;
  die "Expected object\n" if ref $info ne 'HASH';
  my @errors = map "no $_", grep !exists $info->{$_}, qw(keys endpoint);
  push @errors, map "no $_", grep !exists $info->{keys}{$_}, qw(auth p256dh);
  die "Errors found in subscription info: " . join(", ", @errors) . "\n"
    if @errors;
}

1;

=encoding utf8

=head1 NAME

Mojolicious::Plugin::WebPush - plugin to aid real-time web push

=head1 SYNOPSIS

  # Mojolicious::Lite
  my $sw = plugin 'ServiceWorker' => { debug => 1 };
  my $webpush = plugin 'WebPush' => {
    save_endpoint => '/api/savesubs',
    subs_session2user_p => \&subs_session2user_p,
    subs_create_p => \&subs_create_p,
    subs_read_p => \&subs_read_p,
    subs_delete_p => \&subs_delete_p,
    ecc_private_key => 'vapid_private_key.pem',
    claim_sub => "mailto:admin@example.com",
  };

  sub subs_session2user_p {
    my ($c, $session) = @_;
    return Mojo::Promise->reject("Session not logged in") if !$session->{user_id};
    Mojo::Promise->resolve($session->{user_id});
  }

  sub subs_create_p {
    my ($c, $session, $subs_info) = @_;
    app->db->save_subs_p($session->{user_id}, $subs_info);
  }

  sub subs_read_p {
    my ($c, $user_id) = @_;
    app->db->lookup_subs_p($user_id);
  }

  sub subs_delete_p {
    my ($c, $user_id) = @_;
    app->db->delete_subs_p($user_id);
  }

=head1 DESCRIPTION

L<Mojolicious::Plugin::WebPush> is a L<Mojolicious> plugin. In
order to function, your app needs to have first installed
L<Mojolicious::Plugin::ServiceWorker> as shown in the synopsis above.

=head1 METHODS

L<Mojolicious::Plugin::WebPush> inherits all methods from
L<Mojolicious::Plugin> and implements the following new ones.

=head2 register

  my $p = $plugin->register(Mojolicious->new, \%conf);

Register plugin in L<Mojolicious> application, returning the plugin
object. Takes a hash-ref as configuration, see L</OPTIONS> for keys.

=head1 OPTIONS

=head2 save_endpoint

Required. The route to be added to the app for the service worker to
register users for push notification. The handler for that will call
the L</subs_create_p>. If success is indicated, it will return JSON:

  { "data": { "success": true } }

If failure:

  { "errors": [ { "message": "The exception reason" } ] }

This will be handled by the provided service worker. In case it is
required by the app itself, the added route is named C<webpush.save>.

=head2 subs_session2user_p

Required. The code to be called to look up the user currently identified
by this session, which returns a promise of the user ID. Must reject
if no user logged in and that matters. It will be passed parameters:

=over

=item *

The L<Mojolicious::Controller/session> object, to correctly identify
the user.

=back

=head2 subs_create_p

Required. The code to be called to store users registered for push
notifications, which must return a promise of a true value if the
operation succeeds, or reject with a reason. It will be passed parameters:

=over

=item *

The ID to correctly identify the user. Please note that you ought to
allow one person to have several devices with web-push enabled, and to
design accordingly.

=item *

The C<subscription_info> hash-ref, needed to push actual messages.

=back

=head2 subs_read_p

Required. The code to be called to look up a user registered for push
notifications. It will be passed parameters:

=over

=item *

The opaque information your app uses to identify the user.

=back

Returns a promise of the C<subscription_info> hash-ref. Must reject if
not found.

=head2 subs_delete_p

Required. The code to be called to delete up a user registered for push
notifications. It will be passed parameters:

=over

=item *

The opaque information your app uses to identify the user.

=back

Returns a promise of the deletion result. Must reject if not found.

=head2 ecc_private_key

A value to be passed to L<Crypt::PK::ECC/new>: a simple scalar is a
filename, a scalar-ref is the actual key. If not provided,
L</webpush.authorization> will (obviously) not be able to function.

=head2 claim_sub

A value to be used as the C<sub> claim by the L</webpush.authorization>,
which needs it. Must be either an HTTPS or C<mailto:> URL.

=head2 claim_exp_offset

A value to be added to current time, in seconds, in the C<exp> claim
for L</webpush.authorization>. Defaults to 86400 (24 hours). The maximum
valid value in RFC 8292 is 86400.

=head2 push_handler

Override the default push-event handler supplied to
L<Mojolicious::Plugin::ServiceWorker/add_event_listener>. The default
will interpret the message as a JSON object. The key C<title> will be
the notification title, deleted from that object, then the object will be
the options passed to C<< <ServiceWorkerRegistration>.showNotification >>.

See
L<https://developers.google.com/web/fundamentals/push-notifications/handling-messages>
for possibilities.

=head1 HELPERS

=head2 webpush.create_p

  $c->webpush->create_p($user_id, $subs_info)->then(sub {
    $c->render(json => { data => { success => \1 } });
  });

=head2 webpush.read_p

  $c->webpush->read_p($user_id)->then(sub {
    $c->render(text => 'Info: ' . to_json(shift));
  });

=head2 webpush.delete_p

  $c->webpush->delete_p($user_id)->then(sub {
    $c->render(json => { data => { success => \1 } });
  });

=head2 webpush.authorization

  my $header_value = $c->webpush->authorization($subs_info);

Won't function without L</claim_sub> and L</ecc_private_key>, or
C<$subs_info> having a valid URL to get the base of as the C<aud>
claim. Returns a suitable C<Authorization> header value to send to
a push service.  Valid for a period defined by L</claim_exp_offset>.
but could become so to avoid unnecessary computation.

=head2 webpush.public_key

  my $pkey = $c->webpush->public_key;

Gives the app's public VAPID key, calculated from the private key.

=head2 webpush.verify_token

  my $bool = $c->webpush->verify_token($authorization_header_value);

Cryptographically verifies a JSON Web Token (JWT), such as generated
by L</webpush.authorization>.

=head2 webpush.encrypt

  use MIME::Base64 qw(decode_base64url);
  my $ciphertext = $c->webpush->encrypt($data_bytes,
    map decode_base64url($_), @{$subscription_info->{keys}}{qw(p256dh auth)}
  );

Returns the data encrypted according to RFC 8188, for the relevant
subscriber.

=head2 webpush.send_p

  my $result_p = $c->webpush->send_p($jsonable_data, $user_id, $ttl, $urgency);

JSON-encodes the given value, encrypts it according to the given user's
subscription data, adds a VAPID C<Authorization> header, then sends it
to the relevant web-push endpoint.

Returns a promise of the result, which will be a hash-ref with either a
C<data> key indicating success, or an C<errors> key for an array-ref of
hash-refs with a C<message> giving reasons.

If the sending gets a status code of 404 or 410, this indicates the
subscriber has unsubscribed, and L</webpush.delete_p> will be used to
remove the registration. This is considered success.

The C<urgency> must be one of C<very-low>, C<low>, C<normal> (the default)
or C<high>. The C<ttl> defaults to 30 seconds.

=head1 TEMPLATES

Various templates are available for including in the app's templates:

=head2 webpush-askPermission.html.ep

JavaScript functions, also for putting inside a C<script> element:

=over

=item *

askPermission

=item *

subscribeUserToPush

=item *

sendSubscriptionToBackEnd

=back

These each return a promise, and should be chained together:

  <button onclick="
    askPermission().then(subscribeUserToPush).then(sendSubscriptionToBackEnd)
  ">
    Ask permission
  </button>
  <script>
  %= include 'serviceworker-install'
  %= include 'webpush-askPermission'
  </script>

Each application must decide when to ask such permission, bearing in
mind that once permission is refused, it is very difficult for the user
to change such a refusal.

When it is granted, the JavaScript code will communicate with the
application, registering the needed information needed to web-push.

=head1 SEE ALSO

L<Mojolicious>, L<Mojolicious::Guides>, L<https://mojolicious.org>.

L<Mojolicious::Command::webpush> - command-line control of web-push.

RFC 8292 - Voluntary Application Server Identification (for web push).

L<Crypt::RFC8188> - Encrypted Content-Encoding for HTTP (using C<aes128gcm>).

L<https://developers.google.com/web/fundamentals/push-notifications>

=head1 ACKNOWLEDGEMENTS

Part of this code is ported from
L<https://github.com/web-push-libs/pywebpush>.

=cut

__DATA__

@@ webpush-askPermission.html.ep
% # from https://developers.google.com/web/fundamentals/push-notifications/subscribing-a-user
function askPermission() {
  return new Promise(function(resolve, reject) {
    const permissionResult = Notification.requestPermission(resolve);
    if (permissionResult) permissionResult.then(resolve, reject);
  })
  .then(result => result === 'granted' ? result : Promise.reject(result));
}
function subscribeUserToPush() {
  return navigator.serviceWorker.register(
    <%== Mojo::JSON::encode_json(url_for(app->serviceworker->route)) %>
  ).then(registration => registration.pushManager.subscribe({
    userVisibleOnly: true,
    applicationServerKey: urlBase64ToUint8Array(
      <%== Mojo::JSON::encode_json(app->webpush->public_key) %>
    )
  }));
}
function urlBase64ToUint8Array(base64String) {
  const padding = '='.repeat((4 - base64String.length % 4) % 4);
  const base64 = (base64String + padding)
    .replace(/\-/g, '+')
    .replace(/_/g, '/')
  ;
  const rawData = window.atob(base64);
  return Uint8Array.from([...rawData].map((char) => char.charCodeAt(0)));
}
function sendSubscriptionToBackEnd(subscription) {
  return fetch(<%== Mojo::JSON::encode_json(url_for 'webpush.save') %>, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(subscription)
  })
  .then(function(response) {
    if (!response.ok) throw new Error(response.statusText);
    return response.json();
  })
  .then(function(responseData) {
    if (!(responseData.data && responseData.data.success)) {
      throw new Error(responseData.errors[0].message);
    }
  });
}
